File: dotnet/VedaQueryBuilder.cs
using System;
namespace VedaDB;

/// <summary>
/// Fluent query builder for VedaDB.
/// Provides a type-safe, chainable API for constructing SQL queries.
/// </summary>
public class VedaQueryBuilder
{
    private readonly VedaClient _client;
    private readonly string _table;
    private readonly List<string> _columns = new();
    private readonly List<string> _whereClauses = new();
    private readonly List<string> _orderBy = new();
    private readonly List<string> _joins = new();
    private readonly List<string> _groupBy = new();
    private readonly List<string> _having = new();
    private int? _limit;
    private int? _offset;
    private bool _forUpdate;
    private bool _distinct;

    /// <summary>
    /// Create a query builder for a table.
    /// </summary>
    public VedaQueryBuilder(VedaClient client, string table)
    {
        _client = client ?? throw new ArgumentNullException(nameof(client));
        _table = table ?? throw new ArgumentNullException(nameof(table));
    }

    /// <summary>
    /// Select specific columns.
    /// </summary>
    public VedaQueryBuilder Select(params string[] columns)
    {
        _columns.AddRange(columns);
        return this;
    }

    /// <summary>
    /// Add DISTINCT modifier.
    /// </summary>
    public VedaQueryBuilder Distinct()
    {
        _distinct = true;
        return this;
    }

    /// <summary>
    /// Add WHERE clause.
    /// </summary>
    public VedaQueryBuilder Where(string condition)
    {
        _whereClauses.Add($"({condition})");
        return this;
    }

    /// <summary>
    /// Add WHERE column = value.
    /// </summary>
    public VedaQueryBuilder Where(string column, object value)
    {
        _whereClauses.Add($"({column} = {FormatValue(value)})");
        return this;
    }

    /// <summary>
    /// Add WHERE column operator value.
    /// </summary>
    public VedaQueryBuilder Where(string column, string op, object value)
    {
        _whereClauses.Add($"({column} {op} {FormatValue(value)})");
        return this;
    }

    /// <summary>
    /// Add WHERE IN clause.
    /// </summary>
    public VedaQueryBuilder WhereIn(string column, IEnumerable<object> values)
    {
        var vals = string.Join(", ", values.Select(FormatValue));
        _whereClauses.Add($"({column} IN ({vals}))");
        return this;
    }

    /// <summary>
    /// Add WHERE LIKE clause.
    /// </summary>
    public VedaQueryBuilder WhereLike(string column, string pattern)
    {
        _whereClauses.Add($"({column} LIKE '{pattern.Replace("'", "''")}')");
        return this;
    }

    /// <summary>
    /// Add WHERE NULL check.
    /// </summary>
    public VedaQueryBuilder WhereNull(string column)
    {
        _whereClauses.Add($"({column} IS NULL)");
        return this;
    }

    /// <summary>
    /// Add WHERE NOT NULL check.
    /// </summary>
    public VedaQueryBuilder WhereNotNull(string column)
    {
        _whereClauses.Add($"({column} IS NOT NULL)");
        return this;
    }

    /// <summary>
    /// Add ORDER BY clause.
    /// </summary>
    public VedaQueryBuilder OrderBy(string column, bool descending = false)
    {
        _orderBy.Add(descending ? $"{column} DESC" : $"{column} ASC");
        return this;
    }

    /// <summary>
    /// Add LIMIT clause.
    /// </summary>
    public VedaQueryBuilder Limit(int limit)
    {
        _limit = limit;
        return this;
    }

    /// <summary>
    /// Add OFFSET clause.
    /// </summary>
    public VedaQueryBuilder Offset(int offset)
    {
        _offset = offset;
        return this;
    }

    /// <summary>
    /// Add INNER JOIN.
    /// </summary>
    public VedaQueryBuilder Join(string table, string onCondition, string? alias = null)
    {
        var tableRef = alias != null ? $"{table} AS {alias}" : table;
        _joins.Add($"INNER JOIN {tableRef} ON {onCondition}");
        return this;
    }

    /// <summary>
    /// Add LEFT JOIN.
    /// </summary>
    public VedaQueryBuilder LeftJoin(string table, string onCondition, string? alias = null)
    {
        var tableRef = alias != null ? $"{table} AS {alias}" : table;
        _joins.Add($"LEFT JOIN {tableRef} ON {onCondition}");
        return this;
    }

    /// <summary>
    /// Add GROUP BY clause.
    /// </summary>
    public VedaQueryBuilder GroupBy(params string[] columns)
    {
        _groupBy.AddRange(columns);
        return this;
    }

    /// <summary>
    /// Add HAVING clause.
    /// </summary>
    public VedaQueryBuilder Having(string condition)
    {
        _having.Add(condition);
        return this;
    }

    /// <summary>
    /// Add FOR UPDATE clause.
    /// </summary>
    public VedaQueryBuilder ForUpdate()
    {
        _forUpdate = true;
        return this;
    }

    /// <summary>
    /// Build the SQL query string.
    /// </summary>
    public string BuildSql()
    {
        var sb = new System.Text.StringBuilder();

        sb.Append("SELECT ");
        if (_distinct) sb.Append("DISTINCT ");
        sb.Append(_columns.Count > 0 ? string.Join(", ", _columns) : "*");
        sb.Append(" FROM ").Append(_table);

        foreach (var join in _joins)
        {
            sb.Append(" ").Append(join);
        }

        if (_whereClauses.Count > 0)
        {
            sb.Append(" WHERE ").Append(string.Join(" AND ", _whereClauses));
        }

        if (_groupBy.Count > 0)
        {
            sb.Append(" GROUP BY ").Append(string.Join(", ", _groupBy));
        }

        if (_having.Count > 0)
        {
            sb.Append(" HAVING ").Append(string.Join(" AND ", _having));
        }

        if (_orderBy.Count > 0)
        {
            sb.Append(" ORDER BY ").Append(string.Join(", ", _orderBy));
        }

        if (_limit.HasValue)
            sb.Append(" LIMIT ").Append(_limit.Value);

        if (_offset.HasValue)
            sb.Append(" OFFSET ").Append(_offset.Value);

        if (_forUpdate)
            sb.Append(" FOR UPDATE");

        sb.Append(";");
        return sb.ToString();
    }

    /// <summary>
    /// Execute the built query.
    /// </summary>
    public async Task<VedaResult> ExecuteAsync(CancellationToken ct = default)
    {
        var sql = BuildSql();
        VedaMetrics.Increment("vedadb_querybuilder_executions", 1, new() { { "table", _table } });
        return await _client.QueryAsync(sql);
    }

    /// <summary>
    /// Execute and get first row, or null if empty.
    /// </summary>
    public async Task<Dictionary<string, string?>?> FirstAsync(CancellationToken ct = default)
    {
        _limit = 1;
        var result = await ExecuteAsync(ct);
        return result.First();
    }

    /// <summary>
    /// Execute and get a single scalar value.
    /// </summary>
    public async Task<string?> ScalarAsync(CancellationToken ct = default)
    {
        var result = await ExecuteAsync(ct);
        return result.Scalar();
    }

    /// <summary>
    /// Execute and get mapped results.
    /// </summary>
    public async Task<List<T>> MapAsync<T>(Func<Dictionary<string, string?>, T> mapper, CancellationToken ct = default)
    {
        var result = await ExecuteAsync(ct);
        return result.Map(mapper);
    }

    /// <summary>
    /// Execute and get results as dictionaries.
    /// </summary>
    public async Task<List<Dictionary<string, string?>>> ToDictsAsync(CancellationToken ct = default)
    {
        var result = await ExecuteAsync(ct);
        return result.ToDicts();
    }

    /// <summary>
    /// Count rows matching the current filters.
    /// </summary>
    public async Task<long> CountAsync(CancellationToken ct = default)
    {
        var savedColumns = new List<string>(_columns);
        _columns.Clear();
        _columns.Add("COUNT(*)");

        var result = await ExecuteAsync(ct);

        _columns.Clear();
        _columns.AddRange(savedColumns);

        var scalar = result.Scalar();
        return long.TryParse(scalar, out var count) ? count : 0;
    }

    /// <summary>
    /// Check if any rows match the current filters.
    /// </summary>
    public async Task<bool> ExistsAsync(CancellationToken ct = default)
    {
        _limit = 1;
        _columns.Clear();
        _columns.Add("1");

        var result = await ExecuteAsync(ct);
        return result.HasRows;
    }

    private static string FormatValue(object? value) => value switch
    {
        null => "NULL",
        string s => $"'{s.Replace("'", "''")}'",
        bool b => b ? "TRUE" : "FALSE",
        DateTime dt => $"'{dt:yyyy-MM-dd HH:mm:ss}'",
        Guid g => $"'{g}'",
        _ => value.ToString() ?? "NULL"
    };

    /// <summary>
    /// Build an INSERT query for this table.
    /// </summary>
    public async Task<VedaResult> InsertAsync(Dictionary<string, object?> data, CancellationToken ct = default)
    {
        var cols = string.Join(", ", data.Keys);
        var vals = string.Join(", ", data.Values.Select(FormatValue));
        var sql = $"INSERT INTO {_table} ({cols}) VALUES ({vals});";
        return await _client.QueryAsync(sql);
    }

    /// <summary>
    /// Build an UPDATE query from current filters.
    /// </summary>
    public async Task<VedaResult> UpdateAsync(Dictionary<string, object?> set, CancellationToken ct = default)
    {
        var setClause = string.Join(", ", set.Select(kv => $"{kv.Key} = {FormatValue(kv.Value)}"));
        var sb = new System.Text.StringBuilder();
        sb.Append($"UPDATE {_table} SET {setClause}");

        if (_whereClauses.Count > 0)
            sb.Append(" WHERE ").Append(string.Join(" AND ", _whereClauses));

        sb.Append(";");
        return await _client.QueryAsync(sb.ToString());
    }

    /// <summary>
    /// Build a DELETE query from current filters.
    /// </summary>
    public async Task<VedaResult> DeleteAsync(CancellationToken ct = default)
    {
        var sb = new System.Text.StringBuilder();
        sb.Append($"DELETE FROM {_table}");

        if (_whereClauses.Count > 0)
            sb.Append(" WHERE ").Append(string.Join(" AND ", _whereClauses));

        sb.Append(";");
        return await _client.QueryAsync(sb.ToString());
    }
}

File: php/src/VedaQueryBuilder.php
<?php

declare(strict_types=1);

namespace VedaDB;

/**
 * Fluent query builder for VedaDB.
 */
class VedaQueryBuilder
{
    private VedaClient $client;
    private string $table;

    /** @var list<string> */
    private array $columns = ['*'];

    /** @var list<string> */
    private array $wheres = [];

    /** @var list<string> */
    private array $joins = [];

    /** @var list<string> */
    private array $groupBy = [];

    /** @var list<string> */
    private array $having = [];

    /** @var list<string> */
    private array $orderBy = [];

    private ?int $limit = null;
    private ?int $offset = null;
    private bool $distinct = false;
    private bool $forUpdate = false;

    public function __construct(VedaClient $client, string $table)
    {
        $this->client = $client;
        $this->table  = $table;
    }

    /**
     * Set columns to select.
     *
     * @param string ...$columns
     */
    public function select(string ...$columns): self
    {
        $this->columns = empty($columns) ? ['*'] : $columns;
        return $this;
    }

    /**
     * Enable DISTINCT.
     */
    public function distinct(): self
    {
        $this->distinct = true;
        return $this;
    }

    /**
     * Add a WHERE clause.
     */
    public function where(string $column, string $operator, mixed $value): self
    {
        $this->wheres[] = "{$column} {$operator} " . $this->formatValue($value);
        return $this;
    }

    /**
     * Add a WHERE = clause (shorthand).
     */
    public function whereEqual(string $column, mixed $value): self
    {
        return $this->where($column, '=', $value);
    }

    /**
     * Add a WHERE IN clause.
     *
     * @param list<mixed> $values
     */
    public function whereIn(string $column, array $values): self
    {
        $formatted = implode(', ', array_map([$this, 'formatValue'], $values));
        $this->wheres[] = "{$column} IN ({$formatted})";
        return $this;
    }

    /**
     * Add a WHERE IS NULL clause.
     */
    public function whereNull(string $column): self
    {
        $this->wheres[] = "{$column} IS NULL";
        return $this;
    }

    /**
     * Add a WHERE IS NOT NULL clause.
     */
    public function whereNotNull(string $column): self
    {
        $this->wheres[] = "{$column} IS NOT NULL";
        return $this;
    }

    /**
     * Add a raw WHERE condition.
     */
    public function whereRaw(string $condition): self
    {
        $this->wheres[] = $condition;
        return $this;
    }

    /**
     * Add an INNER JOIN.
     */
    public function join(string $table, string $on, ?string $alias = null): self
    {
        $joinTable = $alias ? "{$table} AS {$alias}" : $table;
        $this->joins[] = "INNER JOIN {$joinTable} ON {$on}";
        return $this;
    }

    /**
     * Add a LEFT JOIN.
     */
    public function leftJoin(string $table, string $on, ?string $alias = null): self
    {
        $joinTable = $alias ? "{$table} AS {$alias}" : $table;
        $this->joins[] = "LEFT JOIN {$joinTable} ON {$on}";
        return $this;
    }

    /**
     * Add GROUP BY.
     *
     * @param string ...$columns
     */
    public function groupBy(string ...$columns): self
    {
        $this->groupBy = array_merge($this->groupBy, $columns);
        return $this;
    }

    /**
     * Add HAVING condition.
     */
    public function having(string $condition): self
    {
        $this->having[] = $condition;
        return $this;
    }

    /**
     * Add ORDER BY.
     *
     * @param string ...$columns
     */
    public function orderBy(string ...$columns): self
    {
        $this->orderBy = array_merge($this->orderBy, $columns);
        return $this;
    }

    /**
     * Set LIMIT.
     */
    public function limit(int $limit): self
    {
        $this->limit = max(0, $limit);
        return $this;
    }

    /**
     * Set OFFSET.
     */
    public function offset(int $offset): self
    {
        $this->offset = max(0, $offset);
        return $this;
    }

    /**
     * Enable FOR UPDATE.
     */
    public function forUpdate(): self
    {
        $this->forUpdate = true;
        return $this;
    }

    // -- Execution ----------------------------------------------------------

    /**
     * Execute the SELECT query and return a VedaResult.
     */
    public function get(): VedaResult
    {
        return $this->client->query($this->toSql());
    }

    /**
     * Execute and return all rows as dicts.
     *
     * @return list<array<string, mixed>>
     */
    public function all(): array
    {
        return $this->get()->toDicts();
    }

    /**
     * Execute and return the first row.
     *
     * @return array<string, mixed>|null
     */
    public function first(): ?array
    {
        $this->limit(1);
        return $this->get()->first();
    }

    /**
     * Execute and return a single scalar value.
     */
    public function value(string $column): mixed
    {
        $this->select($column);
        return $this->get()->scalar();
    }

    /**
     * Execute and return a cursor.
     */
    public function cursor(int $fetchSize = 100): VedaCursor
    {
        return $this->client->cursor($this->toSql());
    }

    /**
     * Count rows.
     */
    public function count(): int
    {
        $result = $this->client->query("SELECT COUNT(*) FROM {$this->table}");
        return (int) ($result->scalar() ?? 0);
    }

    /**
     * Check if any rows match.
     */
    public function exists(): bool
    {
        return $this->count() > 0;
    }

    /**
     * Insert a row.
     *
     * @param array<string, mixed> $data
     */
    public function insert(array $data): int
    {
        $cols  = implode(', ', array_keys($data));
        $vals  = implode(', ', array_map([$this, 'formatValue'], array_values($data)));
        $result = $this->client->query("INSERT INTO {$this->table} ({$cols}) VALUES ({$vals});");
        return $result->rowCount;
    }

    /**
     * Insert multiple rows.
     *
     * @param list<array<string, mixed>> $rows
     */
    public function insertMany(array $rows): int
    {
        if (empty($rows)) {
            return 0;
        }
        $cols = implode(', ', array_keys($rows[0]));
        $values = [];
        foreach ($rows as $row) {
            $escaped = array_map([$this, 'formatValue'], array_values($row));
            $values[] = '(' . implode(', ', $escaped) . ')';
        }
        $result = $this->client->query(
            "INSERT INTO {$this->table} ({$cols}) VALUES " . implode(', ', $values) . ';',
        );
        return $result->rowCount;
    }

    /**
     * Update rows.
     *
     * @param array<string, mixed> $data
     */
    public function update(array $data): int
    {
        $setClauses = [];
        foreach ($data as $key => $value) {
            $setClauses[] = "{$key} = " . $this->formatValue($value);
        }
        $sql = "UPDATE {$this->table} SET " . implode(', ', $setClauses);
        if (!empty($this->wheres)) {
            $sql .= ' WHERE ' . implode(' AND ', $this->wheres);
        }
        $result = $this->client->query($sql . ';');
        return $result->rowCount;
    }

    /**
     * Delete rows.
     */
    public function delete(): int
    {
        $sql = "DELETE FROM {$this->table}";
        if (!empty($this->wheres)) {
            $sql .= ' WHERE ' . implode(' AND ', $this->wheres);
        }
        $result = $this->client->query($sql . ';');
        $this->reset();
        return $result->rowCount;
    }

    /**
     * Truncate the table.
     */
    public function truncate(): void
    {
        $this->client->query("TRUNCATE TABLE {$this->table};");
    }

    /**
     * Build the SQL string.
     */
    public function toSql(): string
    {
        $parts = ['SELECT'];

        if ($this->distinct) {
            $parts[] = 'DISTINCT';
        }

        $parts[] = implode(', ', $this->columns);
        $parts[] = "FROM {$this->table}";

        foreach ($this->joins as $join) {
            $parts[] = $join;
        }

        if (!empty($this->wheres)) {
            $parts[] = 'WHERE ' . implode(' AND ', $this->wheres);
        }

        if (!empty($this->groupBy)) {
            $parts[] = 'GROUP BY ' . implode(', ', $this->groupBy);
        }

        if (!empty($this->having)) {
            $parts[] = 'HAVING ' . implode(' AND ', $this->having);
        }

        if (!empty($this->orderBy)) {
            $parts[] = 'ORDER BY ' . implode(', ', $this->orderBy);
        }

        if ($this->limit !== null) {
            $parts[] = "LIMIT {$this->limit}";
        }

        if ($this->offset !== null) {
            $parts[] = "OFFSET {$this->offset}";
        }

        if ($this->forUpdate) {
            $parts[] = 'FOR UPDATE';
        }

        return implode(' ', $parts) . ';';
    }

    /**
     * Reset query state (for reuse).
     */
    public function reset(): self
    {
        $this->columns   = ['*'];
        $this->wheres    = [];
        $this->joins     = [];
        $this->groupBy   = [];
        $this->having    = [];
        $this->orderBy   = [];
        $this->limit     = null;
        $this->offset    = null;
        $this->distinct  = false;
        $this->forUpdate = false;
        return $this;
    }

    private function formatValue(mixed $value): string
    {
        if ($value === null) return 'NULL';
        if (is_bool($value)) return $value ? 'TRUE' : 'FALSE';
        if (is_string($value)) return "'" . str_replace("'", "''", $value) . "'";
        if (is_array($value)) return "'" . str_replace("'", "''", json_encode($value)) . "'";
        return (string) $value;
    }

    public function __toString(): string
    {
        return $this->toSql();
    }
}
